docs/proposals/image-promotion.md

```diff
@@ -0,0 +1,201 @@
+
+```json
+{
+  "type": "ImageChange",
+  "imageChangeParams": {
+    "automatic": true,
+    "from": {
+      "kind": "ImageStreamTag",
+      "name": "origin-ruby-example:prod-ready"
+    },
+    "containerNames": [
+      "frontend"
+    ]
+  }
+}
+```
+
+I also have a BuildConfig defined for my application image that does not specify any triggers.
+
+When a change is detected in remote source code repository, the Jenkins job will connect to OpenShift and
+start a new build.
+Based on the build status, Jenkins job then tag the updated application image as *prod-ready*.
+When a new image with this tag is available, the *frontend-production* manages to roll out a new deployment.
+
+#### Example of using OpenShift
+
+As a user I create two DeploymentConfig's, both as part of my "Application" project:
+
+* *frontend-test*
+* *frontend-production*
+
+The "frontend-test" DeploymentConfig has the ImageChangeTrigger set to watch changes
+in the *qa-ready* ImageStreamTag. This DeploymentConfig also defines the *post* lifecycle
+hook that executes a script in the container deployed and performs additional verification
+and tests.
+
+*Example of DeploymentConfig post-lifecycle hook*:
+```json
+"strategy": {
+  "type": "Rolling",
+  "rollingParams": {
+    "post": {
+      "failurePolicy": "Abort",
+      "execNewPod": {
+        "containerName": "test-container",
+        "command": [
+          "/opt/app-root/bin/verify-deployment"
+        ],
+      }
+    }
+  }
+}
+```
+
+Based on the result of the `verify-deployment` command, which runs in the container based on the
+application image, the command will promote the image by tagging it as *production-ready* (eg. using
+the `oc tag` command). The command might also notify me about the verification failure.
+
+The *frontend-production* DeploymentConfig has the ImageChangeTrigger set to watch changes
+in the *prod-ready* ImageStreamTag. When a new Docker image with this tag is available,
+it can automatically roll out a new production deployment. In case you don't want to roll out
+the *frontend-production* automatically, you can avoid specifying the ImageChangeTrigger and
+roll out the new deployment by hand.
+
+### Manual promotion using external Docker registry
+
+As a developer of a Rails application, when I update the source code and OpenShift
+build my image I want to promote the image by passing it through QA and operation teams.
+
+#### Example
+
+* The image is passed to a remote QA team by pushing the image to a shared Docker registry
+* Remote QA team pull the image and drive the testing and verification
+* Remote QA team push the verified image back to the shared Docker registry, tagged as "verified"
+* Remote devops team pull the image from shared Docker registry and make it available in "stage" cluster
+* Remote devops team promotes the image from "stage" to "production"
+
+### Manual promotion of builder image using external Docker registry
```

Comment:
i would tend to agree that while the implications of the promotion are slightly different (bigger impact), the flow should be similar (maybe the verification looks slightly different or requires more caution, but the basic steps and commands involved should be the same).  i'm not necessarily averse to listing it as a separate use case, but the two use cases shouldn't require separate feature work to support.